xla/service/gpu/runtime/copy_thunk.h

```diff
@@ -18,62 +18,113 @@ limitations under the License.
   uint64_t size_bytes() const { return mem_size_; }
 
- private:
+private:
   const BufferAllocation::Slice source_buffer_;
   const BufferAllocation::Slice destination_buffer_;
   const uint64_t mem_size_;
 };
 
-class DeviceToHostCopyThunk : public DeviceToDeviceCopyThunk {
- public:
-  DeviceToHostCopyThunk(ThunkInfo thunk_info,
-                        const BufferAllocation::Slice& source_buffer,
-                        const BufferAllocation::Slice& destination_buffer,
-                        uint64_t mem_size);
-  absl::Status ExecuteOnStream(const ExecuteParams& params) override;
+//===----------------------------------------------------------------------===//
+// CopyAsyncEvents
+//===----------------------------------------------------------------------===//
+class CopyAsyncEvents {
+public:
+  // Add a new copy-start completion event.
+  // absl::StatusOr<std::shared_ptr<se::Event>> Emplace(
+  absl::Status Emplace(se::StreamExecutor *executor,
+                       const HloInstruction *instr, se::Event &&event);
+
+  // Retrieve a completion event started by copy-start instruction
+  // `instr`, and remove the event from the collection.
+
+  absl::StatusOr<se::Event> Extract(se::StreamExecutor *executor,
+                                    const HloInstruction *instr);
+
+private:
+  using Key = std::pair<se::StreamExecutor *, const HloInstruction *>;
+  absl::Mutex mutex_;
+  absl::flat_hash_map<Key, se::Event> events_ ABSL_GUARDED_BY(mutex_);
 };
 
-class HostToDeviceCopyThunk : public DeviceToDeviceCopyThunk {
- public:
-  HostToDeviceCopyThunk(ThunkInfo thunk_info,
-                        const BufferAllocation::Slice& source_buffer,
-                        const BufferAllocation::Slice& destination_buffer,
-                        uint64_t mem_size);
-  absl::Status ExecuteOnStream(const ExecuteParams& params) override;
+//===----------------------------------------------------------------------===//
+// DeviceHostCopyThunk
+//===----------------------------------------------------------------------===//
+// A thunk that copies data from a device buffer to a host buffer.
+class DeviceHostCopyThunk : public DeviceToDeviceCopyThunk {
+public:
+  // Constructs a CopyThunk that copies host data from `source_buffer` to the
+  // device buffer `destination_buffer`. `mem_size` is the size of the data in
+  // bytes. `instr` is the copy-start instruction. `device_to_host` indicates
+  // whether to generate device-to-host or host-to-device memcpy.
+  DeviceHostCopyThunk(ThunkInfo thunk_info,
+                      const BufferAllocation::Slice &source_buffer,
+                      const BufferAllocation::Slice &destination_buffer,
+                      uint64_t mem_size,
+                      std::shared_ptr<CopyAsyncEvents> events,
+                      const HloInstruction *instr, bool device_to_host);
+  absl::Status ExecuteOnStream(const ExecuteParams &params) override;
+
+private:
+  const ExecutionStreamId stream_id_;
```

Comment:
like above, base Thunk class already has this field.

Reply:
Will remove